### PR TITLE
Assignment7+8 - Joshua Merrell and Sergei Ashikhmin

### PR DESCRIPTION
**Pull request by joshmerrell**
@joshmerrell and @SergeiAshikhmin

### CODE REVIEW

**Comment by Purplepigs0129** [comment]
Reviewing :)

File: Assignment/SampleData.cs [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+    /// does not keep the stream open as an object, the IDisposable pattern is not needed
+    /// </summary>
+    /// <param name="path">path to a CSV-formated file</param>
+    /// <returns>an object containing the lines</returns>
+    public static SampleData FromFile(string path) =>
+        FromCsvString(File.ReadAllText(path));
+
+    // 2.
+    public IEnumerable<string> GetUniqueSortedListOfStatesGivenCsvRows() =>
+        (from person in People
```
Comment (Purplepigs0129) [inline Assignment/SampleData.cs:59]: Interesting approach to use Csv Rows. Great Job!

File: Assignment/SampleData.cs [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+    /// Returns a sorted comma separated list of unique states. Doing it this way instead
+    /// of calling ToArray on the result of GetUniqueSortedListOfStatesGivenCsvRows()
+    /// allows us to make use of lazy evaluation 
+    /// </summary>
+    /// <returns></returns>
+    public string GetAggregateSortedListOfStatesUsingCsvRows()
+    {
+        //return string.Join(",", GetUniqueSortedListOfStatesGivenCsvRows().ToArray());
+        return GetUniqueSortedListOfStatesGivenCsvRows()
+            .Aggregate((stateL, stateR) => $"{stateL},{stateR}");
```
Comment (Purplepigs0129) [inline Assignment/SampleData.cs:73]: I can see you are selecting states but no ToArray() could be okay because it says consider and not that it is an actual requirement to use it but marking it wrong as that is what it says

File: Directory.Build.props [hunk 3]
```diff
@@ -1,6 +1,6 @@
 <Project>
 <PropertyGroup>
-    <TargetFramework>net6.0</TargetFramework>
+    <TargetFramework>net8.0</TargetFramework>
```
Comment (Purplepigs0129) [inline Directory.Build.props:3]: I believe you are missing the latest LangVersion in both this and CSPROJ files

File: Assignment.Tests/TestingCsvData.cs [hunk 4]
```diff
@@ -0,0 +1,77 @@
+namespace Assignment.Tests;
+
+public static class TestingCsvData
+{
+
+    public const string InconsistentColumnData = @"A,B,C,D
```
Comment (Purplepigs0129) [inline Assignment.Tests/TestingCsvData.cs:6]: Great Job testing CSV data would have never thought of it since I figure the other tests we did kind of confirmed that we were using the right file.

File: Assignment/Person.cs [hunk 5]
```diff
@@ -0,0 +1,48 @@
+﻿using System.Linq;
+using System.Collections.Generic;
+using System;
+
+namespace Assignment;
+public class Person(string firstName, string lastName, IAddress address, string emailAddress) : IPerson
+{
+    public static Person FromCsvRow(string row)
```
Comment (Purplepigs0129) [inline Assignment/Person.cs:8]: Interesting to use this for the Person method in SampleData instead of just doing it in the SampleData Great Work!

File: Assignment.Tests/PersonTests.cs [hunk 6]
```diff
@@ -0,0 +1,58 @@
+﻿using System;
+
+namespace Assignment.Tests;
+
+[TestClass]
+public class PersonTests
+{
```
Comment (Purplepigs0129) [inline Assignment.Tests/PersonTests.cs:7]: Great Tests for the file you introduced to complete this assignment

**Review by Purplepigs0129 — COMMENTED** [review]
Implement the ISampleData.CsvRows property, loading the data from the People.csv file and returning each line as a single string. ✔

Change the "Copy to" property on People.csv to "Copy if newer" so that the file is deployed along with your test project. ✔
Using LINQ, skip the first row in the People.csv. ❌\✔
Be sure to appropriately handle resource (IDisposable) items correctly if applicable (and it may not be depending on how you implement it). ✔
Implement IEnumerable<string> GetUniqueSortedListOfStatesGivenCsvRows() to return a sorted, unique list of states. ✔

Use ISampleData.CsvRows for your data source. ✔
Don't forget the list should be unique. ✔
Sort the list alphabetically. ✔
Include a test that leverages a hardcoded list of addresses. ✔
Include a test that uses LINQ to verify the data is sorted correctly (do not use a hardcoded list). ✔
Implement ISampleData.GetAggregateSortedListOfStatesUsingCsvRows() to return a string that contains a unique, comma separated list of states. ✔

Use ISampleData.GetUniqueSortedListOfStatesGivenCsvRows() for your data source. ✔
Consider "selecting" only the states and calling ToArray() to retrieve an array of all the state names. ❌
Given the array, consider using string.Join to combine the list into a single string. ❌✔
Implement the ISampleData.People property to return all the items in People.csv as Person objects ✔

Use ISampleData.CsvRows as the source of the data. ✔
Sort the list by State, City, and Zip. (Sort the addresses first then select). ✔
Be sure that Person.Address is also populated. ✔
Adding null validation to all the Person and Address properties is optional.
Consider using ISampleData.CsvRows in your test to verify your results. ✔
Implement ISampleDate.FilterByEmailAddress(Predicate<string> filter) to return a list of names where the email address matches the filter. ✔

Use ISampleData.People for your data source. ✔
Implement ISampleData.GetAggregateListOfStatesGivenPeopleCollection(IEnumerable<IPerson> people) to return a string that contains a unique, comma-separated list of states. ✔

Use the people parameter from ISampleData.People property for your data source. ✔
At a minimum, use the System.Linq.Enumerable.Aggregate` LINQ method to create your result. ✔
Don't forget the list should be unique. ✔
It is recommended that, at a minimum, you use ISampleData.GetUniqueSortedListOfStatesGivenCsvRows to validate your result.
Given the implementation of Node in Assignment5

Implement IEnumerable<T> to return all the items in the "circle" of items. ✔
Add an IEnumberable<T> ChildItems(int maximum) method to Node that returns the remaining items with a maximum number of items returned less than maximum.
Extra Credit
Implement the homework using async/await and multi-threading by defining a new SampleDataAsync class that implements IAsyncSampleData). Refactor your SampleData and SampleDataAsync classes with minimal duplication. Be sure to refactor your tests to re-use a significant amount of the test code for both implementations. ❌
Fundamentals
Place all shared project properties into a Directory.Build.Props file.
Place all shared project items into a Directory.Build.targets file.
Ensure nullable reference types is enabled ✔
Ensure that you turn on code analysis for all projects(EnableNETAnalyzers) ✔
Set LangVersion and the TargetFramework to the latest released versions available (preview versions optional) ❌
and enabled .NET analyzers for both projects ✔
For this assignment, consider using Assert.AreEqual<T>() (the generic version) ✔
All of the above should be unit tested ✔
Choose simplicity over complexity ✔

Great Work here only minor thing was the ToArray() but I don't think you need it as it said consider not to actually do it that way. The way you are doing it seems fine. Otherwise I found it interesting that you guys decided to introduce a Person file but it actually makes a lot of sense now. Good Job!

**Comment by Tvor369** [comment]
Also reviewing

File: Assignment/SampleData.cs [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+    public int Columns { get; init; }
+
+    /// <summary>
+    /// FromCsvString validates CSV-formated data and returns a container object
+    /// </summary>
+    /// <param name="csv">a CSV string</param>
+    /// <returns>an object containing the lines</returns>
+    /// <exception cref="Exception"></exception>
+    public static SampleData FromCsvString(string csv) {
+        ArgumentException.ThrowIfNullOrEmpty(csv);
```
Comment (Tvor369) [inline Assignment/SampleData.cs:23]: Very interesting take on the first part of the assignment, I didn't even think about checking if the csv file had consistent columns throughout, makes sense though since that would really mess up all the query methods

File: Assignment/SampleData.cs [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+
+        IEnumerable<IPerson> people = People.Where(person => filter(person.EmailAddress));
+
+        return people.Select(person => (person.FirstName, person.LastName));
+    }
+
+    // 6.
+    public string GetAggregateListOfStatesGivenPeopleCollection(IEnumerable<IPerson> people)
+    {
+        ArgumentNullException.ThrowIfNull(people);
```
Comment (Tvor369) [inline Assignment/SampleData.cs:103]: I think you are missing a test for if you pass in a null on this one

File: Assignment/NodeEnumerator.cs [hunk 9]
```diff
@@ -0,0 +1,42 @@
+﻿using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace Assignment;
+
```
Comment (Tvor369) [inline Assignment/NodeEnumerator.cs:7]: All I did was just implement the methods it auto generates when you make Node implement IEnumerable

File: Assignment.Tests/NodeTests.cs [hunk 10]
```diff
@@ -0,0 +1,77 @@
...
+        node.Append(2);
+        node.Append(3);
+        node.Append(4);
+        Assert.AreEqual(1, node.Value);
+        Assert.AreEqual(2, node.Next.Value);
+        Assert.AreEqual(3, node.Next.Next.Value);
+        Assert.AreEqual(4, node.Next.Next.Next.Value);
+    }
+
+    [TestMethod]
```
Comment (Tvor369) [inline Assignment.Tests/NodeTests.cs:48]: Good job testing that Node can now work with Linq

**Comment by Tvor369** [comment]
### Instructions

1. Implement the ISampleData.CsvRows property, loading the data from the People.csv file and returning each line as a single string. ✔
- Change the "Copy to" property on People.csv to "Copy if newer" so that the file is deployed along with your test project. ✔
- Using LINQ, skip the first row in the People.csv. ✔
- Be sure to appropriately handle resource (IDisposable) items correctly if applicable (and it may not be depending on how you implement it). ✔
2. Implement IEnumerable<string> GetUniqueSortedListOfStatesGivenCsvRows() to return a sorted, unique list of states. ✔
- Use ISampleData.CsvRows for your data source. ✔
- Don't forget the list should be unique. ✔
- Sort the list alphabetically. ✔
- Include a test that leverages a hardcoded list of addresses. ✔
- Include a test that uses LINQ to verify the data is sorted correctly (do not use a hardcoded list). ✔

3. Implement ISampleData.GetAggregateSortedListOfStatesUsingCsvRows() to return a string that contains a unique, comma separated list of states. ✔
- Use ISampleData.GetUniqueSortedListOfStatesGivenCsvRows() for your data source. ✔
- Consider "selecting" only the states and calling ToArray() to retrieve an array of all the state names. ✔
- Given the array, consider using string.Join to combine the list into a single string. ✔

4. Implement the ISampleData.People property to return all the items in People.csv as Person objects ✔

- Use ISampleData.CsvRows as the source of the data. ✔
- Sort the list by State, City, and Zip. (Sort the addresses first then select). ✔
- Be sure that Person.Address is also populated. ✔
- Adding null validation to all the Person and Address properties is optional.
- Consider using ISampleData.CsvRows in your test to verify your results. ✔

5. Implement ISampleDate.FilterByEmailAddress(Predicate<string> filter) to return a list of names where the email address matches the filter. ✔

- Use ISampleData.People for your data source. ✔

6. Implement ISampleData.GetAggregateListOfStatesGivenPeopleCollection(IEnumerable<IPerson> people) to return a string that contains a unique, comma-separated list of states. ✔

- Use the people parameter from ISampleData.People property for your data source. ✔
- At a minimum, use the System.Linq.Enumerable.Aggregate` LINQ method to create your result. ✔
- Don't forget the list should be unique. ✔
- It is recommended that, at a minimum, you use ISampleData.GetUniqueSortedListOfStatesGivenCsvRows to validate your result.
- Given the implementation of Node in Assignment5

7. Implement IEnumerable<T> to return all the items in the "circle" of items. ✔
- Add an IEnumberable<T> ChildItems(int maximum) method to Node that returns the remaining items with a maximum number of items returned less than maximum.✔

### Fundamentals
- Place all shared project properties into a Directory.Build.Props file.✔
- Place all shared project items into a Directory.Build.targets file.
- Ensure nullable reference types is enabled ✔
- Ensure that you turn on code analysis for all projects(EnableNETAnalyzers) ✔
- Set LangVersion and the TargetFramework to the latest released versions available (preview versions optional) ❌ ←Missing LangVersion
- and enabled .NET analyzers for both projects ✔
- For this assignment, consider using Assert.AreEqual<T>() (the generic version) ✔
- All of the above should be unit tested ✔
- Choose simplicity over complexity ✔ ← CsvRows was more complicated than I expected but it made sense to check for the column consistency

File: Assignment.Tests/NodeTests.cs (hunk 10 above)
Comment (SergeiAshikhmin) [reply Assignment.Tests/NodeTests.cs:48]: Thanks

File: Assignment/NodeEnumerator.cs (hunk 9 above)
Comment (SergeiAshikhmin) [reply Assignment/NodeEnumerator.cs:7]: I agree with you, but this code was done by Josh and I think he wanted to experiment with these interfaces.

File: Assignment/SampleData.cs (hunk 2 above)
Comment (joshmerrell) [reply Assignment/SampleData.cs:73]: I originally did that, but decided to use Aggregate to join them instead, because I figured that "consider" meant that it wasn't a hard requirement as long as we justified our reason for not doing it. Doing it the original way (the commented out code) requires turning the IEnumerable into an array and then go through every element to join them into a string, which would be less optimized since it requires each element to be processed more times than using Aggregate()

File: Assignment/Person.cs [hunk 11]
```diff
@@ -0,0 +1,48 @@
...
+{
+    public static Person FromCsvRow(string row)
+    {
+        ArgumentException.ThrowIfNullOrEmpty(row);
+        string[] columns = row.Split(',');
+        if(columns.Length != 8)
+        {
+            throw new ArgumentException($"Row requires 8 columns, got {columns.Length} columns");
+        }
+        return new(columns[1], columns[2],
```
Comment (BenjaminMichaelis) [inline Assignment/Person.cs:16]: for specific columns in this file, a class or enum would be nice to give readable values instead of like '6' or '2' to different columns, so you could say like `[Column.State]` or `[Column.City]`, etc.

File: Assignment/Person.cs [hunk 12]
```diff
@@ -0,0 +1,48 @@
...
+            columns[3]);
+    }
+
+    public void Deconstruct(out string FirstName, out string LastName, out string EmailAddress,
+        out string StreetAddress, out string City, out string State, out string Zip)
+    {
+        FirstName = this.FirstName;
+        LastName = this.LastName;
+        EmailAddress = this.EmailAddress;
+        (StreetAddress, City, State, Zip) = (Address)Address;
```
Comment (BenjaminMichaelis) [inline Assignment/Person.cs:27]: having to use `this` should be an indication that the parameters are not named properly, they should be camelCase

File: Assignment/Node.cs [hunk 13]
```diff
@@ -0,0 +1,84 @@
...
+        return atEnd;
+    }
+
+    public IEnumerable<Node<T>> ChildItems(int maximum)
+    {
+        int n = 0;
+        return Next.TakeWhile(child => n++ < maximum);
+    }
+
+    public IEnumerator<Node<T>> GetEnumerator() => new NodeEnumerator<T>(this);
```
Comment (BenjaminMichaelis) [inline Assignment/Node.cs:81]: Why do you need a NodeEnumerator<T>
